misk/src/main/kotlin/misk/clustering/kubernetes/KubernetesClusterService.kt

```diff
@@ -0,0 +1,207 @@
+    if (!changes.hasDiffs) return
+
+    peersRef.set(changes.current)
+    hashRingRef.set(ClusterHashRing(changes.current))
+
+    watches.forEach {
+      try {
+        it(changes)
+      } catch (th: Throwable) {
+        log.error(th) { "error triggering watch in response to cluster change" }
+      }
+    }
+  }
+
+  /**
+   * Handles a request for a new watch, adding it to the watch set and triggering it with
+   * the current set of peers
+   */
+  private fun handleNewWatch(
+    newWatch: ClusterWatch,
+    peers: Set<Cluster.Member>,
+    watches: MutableSet<ClusterWatch>
+  ) {
+    watches.add(newWatch)
+    try {
+      newWatch(Cluster.Changes(peers))
+    } catch (th: Throwable) {
+      log.error(th) { "error triggering watch during registration" }
+    }
+  }
+
+  /** Applies an incoming change to a set of cluster members */
+  private fun applyChange(
+    peers: MutableMap<String, Cluster.Member>,
+    changeType: String,
+    pod: V1Pod
+  ): Cluster.Changes {
+    val member = Cluster.Member(pod.metadata.name, pod.status.podIP ?: "")
+
+    // Don't bother to include ourselves in the peer list
+    if (member.name == self.name) return Cluster.Changes(peers.values.toSet())
+
+    return when (changeType) {
+      CHANGE_TYPE_ADDED, CHANGE_TYPE_MODIFIED -> {
+        val isReady = pod.status.containerStatuses?.all { it.isReady } ?: false
+        if (isReady && member.ipAddress.isNotBlank()) {
+          log.info { "added peer ${member.name}" }
+          peers[member.name] = member
+          Cluster.Changes(peers.values.toSet(), added = setOf(member))
+        } else if (peers.containsKey(member.name)) {
+          log.info { "peer ${member.name} no longer ready" }
+          peers.remove(member.name)
+          Cluster.Changes(peers.values.toSet(), removed = setOf(member))
+        } else Cluster.Changes(peers.values.toSet())
+      }
+
+      CHANGE_TYPE_DELETED -> {
+        if (peers.containsKey(member.name)) {
+          log.info { "deleted peer ${member.name}" }
+          peers.remove(member.name)
+          Cluster.Changes(peers.values.toSet(), removed = setOf(member))
+        } else Cluster.Changes(peers.values.toSet())
+      }
+      else -> Cluster.Changes(peers.values.toSet())
+    }
+  }
+
+  /**
+   * An internal event handled by the cluster service. All interactions with the cluster service
+   * are dispatched as events polled by the event loop]
+   */
+  data class Event(
+    val type: Type,
+    val watchResponse: Watch.Response<V1Pod>? = null,
+    val newWatch: ClusterWatch? = null,
+    val syncPointCallback: (() -> Unit)? = null
+  ) {
+    enum class Type {
+      // Sent when a cluster changes, contains a watch response
+      CLUSTER_CHANGE,
+
+      // Sent when a new watch is registered, contains a ClusterWatch
+      NEW_WATCH,
+
+      // Sent when a sync point is requested, contains a callback triggered when the event is processed
+      SYNC_POINT,
```

Comment:
for my own knowledge, when is a sync point requested? not following the use case

Reply:
It's a way of saying - call me back when all of the events currently on the queue have been processed. It's mostly for testing. I couldn't think of a good name so if you have a better one I'd be open to it.

Reply:
ah. well maybe just include that in the comment